### ISSUE TITLE
Question Set 4

### ISSUE THREAD

**Issue by andrewniehaus** [andrewniehaus/adleoAN#4]
---
Question Set 4
---

The following questions relate to readings and practical materials up through the second section on [Semantic Segmentation](../semseg2.qmd). They are worth 20 points total. 

Submit the answers to these questions by placing a GitHub issue entitled "Question Set 4" on your assignments repo. Provide the answers to each question in your issue. 

### Q1

a. Can you explain the limitations or drawbacks associated with each of the following methods used to extend the receptive field for deep learning in semantic segmentation:

- Using a larger kernel size for conv layers throughout the network.

*Larger kernels have potential to lose sensitivity to smaller features.*

- Increasing the depth of the network by stacking more convolutional layers.

*Increasing the depth of the network risks overfitting.*

- Downsampling using pooling or convolution with a stride greater than 1.

*Using a stride greater than 1 risks difficulty in specifically detecting object edges/boundaries*

- Using dilated convolutions.

*Dilated convolutions risk gridded artifacts*

b.  What are the proposed solutions to the drawbacks of using dilated conv layers?

*Gridded artifacts can be avioded with ASPP or Pyramid Pooling*

### Q2

Compare and contrast the usage of parallel and serial arrangements of dilated convolution layers for increasing the receptive field in a deep neural network. How does the arrangement affect the size and shape of the receptive field, and what are some benefits and drawbacks of each approach?

*Parallel and serial dilation differ in the sequencing of how the dilation happens. Serial dilation has a greater risk of exploding/vanishing gradients but is better for gettinga larger receptive field. Parallel dilation is more computationally efficient but are less able to get really deep.*

**Hint:** There are two main ways to arrange dilated convolution layers in a network: in parallel or in series. In the parallel arrangement, multiple dilated convolutions are applied in parallel to the same input feature map, and their outputs are concatenated. In the serial arrangement, multiple dilated convolutions are applied in sequence, with the output of one layer serving as the input to the next layer. The answers to this question are available in the reading.

**Comment by Rahebe22** [comment]
Thanks for the answers!

Please also check below:

Q1 part a:
The limitations or drawbacks associated with each of the following methods used to extend the receptive field for deep learning in semantic segmentation:

Using a larger kernel size for conv layers throughout the network: Increasin the kernal size will definitely create mor elearnable parameters which will increase computation time and cost but it depends on the image as well as the task. It can also lead to overfitting.
Increasing the depth of the network by stacking more convolutional layers: The problem of vanishing gradients start to appear if this is done and it also increases computation time and cost.
Downsampling using pooling or convolution with a stride greater than 1: This depends on the task, but if we doing something like boundary or edge detection then there is loss in spatial detail. Even though we can upsample it is very hard to regain the precision that we lose.
Using dilated convolutions: This gives rise to the "checkerboard" effect as it skips over pixels and the outputs appear blocky or disconnected and we also lose the fine details.

Q1 part b:

When dilated convolutions are employed, spacially in a sequence with increasing dilation rates, they can create undesired grid-like patterns and artifacts in the output feature maps. This is because the adjacent units in the output layer have independent effective receptive fields so the number of layers in the sequence rises, and as a result, the number of neighboring units with non-overlapping receptive fields also increases. Using rising dilation rates by a common factor in the successive layers of the sequence can worsen this problem, as it guarantees that the effective receptive field of the neighboring units in the succeeding activation maps do not overlap. The issue can also become more severe if large dilation rates are utilized since the distance between non-zero elements in the filter becomes excessively large, leading to feature maps that are inconsistent. Furthermore, if multiple dilated convolutions with the same rate are applied in a sequence, a pixel in the top layer may lose local information as it is only influenced by a small fraction of the pixels in its effective receptive field (ERF) from earlier layers. This can create an output with a checkerboard pattern.
there are several methods proposed to deal with this issue: ESP method, using a merging strategy that involves merging the feature maps from a branch with a lower dilation rate to the neighboring branch that has a higher dilation rate, offers a resolution to the problem of gridding when layers are arranged in parallel. Hybrid Dilated Convolution separates dilated convolutions into sets of three and assigns different dilation rates to each convolution within a set. A-ASPP (augmented-ASPP) module, employs a series of 3x3 dilated convolution branches that alternate between increasing and decreasing dilation rates. This approach leads to a denser sampling and reduces the gridding problem in the output feature maps. The Large Kernel Pyramid Pooling module (LKPP) also aims to detect objects at different scales by fusing local and long-range dependencies. (Khallaghi et al.)

Q2

Parallel and serial arrangements of dilated convolution layers are two common methods for increasing the receptive field in a deep neural network. The parallel and cascaded arrangements of dilated convolution layers have different effects on the size and shape of the receptive field in a deep neural network. In the parallel arrangement, multiple branches of dilated convolution layers with different dilation rates are applied in parallel, and the resulting feature maps are concatenated to form the output. This approach can capture a large receptive field by combining information from multiple branches, which is useful for detecting irregularly shaped objects in indoor scenes however this architecture may not be precise when there is class imbalance. On the other hand, in the cascaded arrangement, multiple layers with increasing dilation rates are applied sequentially, which expands the receptive fields more than parallel design. This approach can better capture the global context of the image.
